[JEJE] Step2 - 상속, 프로토콜, 객체간 통신

jung-yun opened the pull request:
## 주요 작업 목록
- [x] 자판기 구조체 설계
- [x] 자판기 금액을 원하는 금액만큼 올리는 메소드
- [x] 특정 상품 인스턴스를 넘겨서 재고를 추가하는 메소드
- [x] 현재 금액으로 구매가능한 음료수 목록을 리턴하는 메소드
- [x] 음료수를 구매하는 메소드
- [x] 잔액을 확인하는 메소드
- [ ] 전체 상품 재고를 (사전으로 표현하는) 종류별로 리턴하는 메소드
- [x] 유통기한이 지난 재고만 리턴하는 메소드
- [ ] 따뜻한 음료만 리턴하는 메소드
- [ ] 시작이후 구매 상품 이력을 배열로 리턴하는 메소드
- [ ] 각 메소드 동작을 검증할 수 있는 테스트 함수를 작성
- [x] 전체 코드를 확인할 수 있는 통합 테스트 시나리오를 가지고 동작을 확인
- [x] 상-하위 클래스 구현 및 이니셜라이저 활용
- [x] 클래스에 프로토콜 적용. 동일 프로토콜 채용한 객체간 통신
- [ ] 클로저와 핸들러를 이용한 객체간 통신 공부와 해당 메소드 구현
- [ ]  XCTest 활용
- [x] Date 자료형 비교 

## 학습 키워드
'프로토콜', '상속', '클로저', '핸들러', 'XCTest', '인터페이스'

## 고민과 해결
1. (be28da6)Machine 구조체가 포함하는 두 클래스 간의 통신을 프로토콜로 구현하고 싶었습니다. 프로토콜 사용 방법에 대해 구글링해봐도 잘 이해가 되지 않아 스스로 여러가지 시도를 해봤습니다. 그리고 두 객체가 동일한 프로토콜을 차용하고 있으면 공유하는 프로퍼티와 메소드로 두 객체간에 통신이 가능할 거란 생각이 들었습니다.
```
protocol MoneyExchangable {
    var money: Int { get set }
    
    func increaseMoney(by amount: Int)
    func deductMoney(by amount: Int)
    func sendMoney(amount: Int, to counterPart: MoneyExchangable)
}
```

```
`class MoneyStorage: MoneyExchangable {
    var money: Int
    // ...
    func increaseMoney(by amount: Int) {
        money += amount
    }
    
    func deductMoney(by amount: Int) {
        money -= amount
    }
    // ...
    func sendMoney(amount: Int, to counterPart: MoneyExchangable) {
        self.deductMoney(by: amount)
        counterPart.increaseMoney(by: amount)
    }
}
```

## 질문거리
1. (키워드 발견) 계층구조의 클래스에 프로토콜 인터페이스 적용.
유통기한 확인 `validate(with date:Date) -> Bool`
뜨거운 음료여부 `isHot()`
저칼로리 여부 `isLowCalorie() -> Bool`
등의 함수를 한 프로토콜에 넣어 그 프로토콜을 계층구조 상 최상위 클래스인 Beverage 클래스가 차용했을때 isHot 혹은 isLowCalorie를 판단할 수 있는 프로퍼티가 없는 하위 클래스들이 있다는 것을 알아차렸습니다. 프로토콜을 객체마다 만들자니 좀 더 쉬운 방법이 있을까 고민하게 됐습니다. protocol extension을 알아봤는데 프로토콜에 default 값이 있는 메소드를 추가할 때 사용하는 것 같아 적용하지 않았습니다. 이 경우 어떻게 해야하는 지 궁금합니다. 다음 시도해 볼 방법으론 계층구조 상 최하위 클래스마다 해당 프로토콜을 작성하는 것입니다. 키워드: **Fat interface**

#### 이번 스텝을 진행하면서 아쉬웠던 점...
여러 클래스를 동시에 만지다 보니 commit 단위가 깔끔하지 않고 뒤죽박죽인 것 같습니다..

jung-yun committed:
Delete format() which is irrelevant to Beverage class and move the functionality into date datatype

jung-yun committed:
Implement subclasses

jung-yun committed:
Add protocol to Beverage class

jung-yun committed:
Create MoneyStorage class and Machine structure

Deleted obsolete test codes for previous commit
Add methods checking and adding money in moneyStorage class

jung-yun committed:
Create addStock method and relevant ones

jung-yun committed:
Create methods to show currently purchasable beverage

jung-yun committed:
Tried another method to show desired beverage list

jung-yun committed:
Implement beverage purchase behavior

Added MoneyExchangable protocol to utilize money change between MoneyStorage and BeverageStorage
Edited relevant methods and fixed ambiguous names
Activated purchaseBeverage method in another thread to test if running the method not in main run loop cause trouble

jung-yun committed:
Implement transaction stop button

jung-yun committed:
Create Item class

Fixed the relevant method- addStock

jung-yun committed:
Refactor stock showing methods

jung-yun committed:
Implement showExpired method

showExpired method shows expired stocks only
Also edited initializers in subclass to activate manufactuingDate data setting

jung-yun committed:
Delete unwanted methods in protocol

jung-yun committed:
Add test scenario

Refactor code to print result:more cleary in console window
Edit relevant methods' logic

godrm commented:
> ## 주요 작업 목록
> * [x]  자판기 구조체 설계
> * [x]  자판기 금액을 원하는 금액만큼 올리는 메소드
> * [x]  특정 상품 인스턴스를 넘겨서 재고를 추가하는 메소드
> * [x]  현재 금액으로 구매가능한 음료수 목록을 리턴하는 메소드
> * [x]  음료수를 구매하는 메소드
> * [x]  잔액을 확인하는 메소드
> * [ ]  전체 상품 재고를 (사전으로 표현하는) 종류별로 리턴하는 메소드
> * [x]  유통기한이 지난 재고만 리턴하는 메소드
> * [ ]  따뜻한 음료만 리턴하는 메소드
> * [ ]  시작이후 구매 상품 이력을 배열로 리턴하는 메소드
> * [ ]  각 메소드 동작을 검증할 수 있는 테스트 함수를 작성
> * [x]  전체 코드를 확인할 수 있는 통합 테스트 시나리오를 가지고 동작을 확인
> * [x]  상-하위 클래스 구현 및 이니셜라이저 활용
> * [x]  클래스에 프로토콜 적용. 동일 프로토콜 채용한 객체간 통신
> * [ ]  클로저와 핸들러를 이용한 객체간 통신 공부와 해당 메소드 구현
> * [ ]   XCTest 활용
> * [x]  Date 자료형 비교

완료가 안된 기능은 일부러 빠드린 걸까요?

godrm commented on VendingMachineApp/VendingMachineApp/BeverageModel/Beverage.swift:
```diff
@@ -0,0 +1,64 @@
+    }
+}
+
+extension Date {
+    func useSimpleFormat() -> String {
+        let format = DateFormatter()
+        format.dateFormat = "yyyyMMdd"
+        return format.string(from: self)
+    }
+}
```
음료와 Date는 직접적인 관련이 없으니까 파일을 분리해도 되겠습니다. 
format을 매개변수로 넘길 수 있으면 좀 더 유연한 메소드가 되지 않을까요?

godrm commented on VendingMachineApp/VendingMachineApp/BeverageModel/Beverage.swift:
```diff
@@ -0,0 +1,64 @@
+    }
+}
+
+extension Beverage: Drinkable {
+    func isExpired(by date: Date) -> Bool {
+        return date > packDate.addingTimeInterval(86400 * 60)
+    }
+}
```
이런 프로토콜을 지원하는 것은 좋네요.
`86400 * 60` 에서 86400, 60 같은 숫자도 어떤 의미인지 상수로 표현해보세요.

godrm commented on VendingMachineApp/VendingMachineApp/BeverageModel/Beverage.swift:
```diff
@@ -0,0 +1,64 @@
+            return true
+        } else {
+            return false
+        }
```
이제 이런 메소드는 Equatable 프로토콜로 처리하는 것도 좋습니다. 
비교 조건이 있을 때는 return `비교조건` 형태도 깔끔합니다.

godrm commented on VendingMachineApp/VendingMachineApp/BeverageModel/Beverage.swift:
```diff
@@ -0,0 +1,64 @@
+    private let packDate: Date
+    
+    public init(brandName: String, sizeInMilliliter: Int, itemPrice: Int, itemName: String, manufacturingDate: Date) {
+        brand = brandName
+        size = sizeInMilliliter
+        price = itemPrice
+        name = itemName
+        packDate = manufacturingDate
```
매개변수 이름이랑 내부 속성 이름을 다르게 한 것은 일부러 그렇게 한걸까요?
오히려 name이 반복되고, 혼란스럽지는 않은가요? 물론 꼭 같아야 하는 건 아닙니다.
예를 들어 sizeInMilliliter처럼 단위를 이름에 포함하기 보다, Int 대신 타입을 지정하는 게 더 좋습니다.

godrm commented on VendingMachineApp/VendingMachineApp/BeverageModel/CoffeeModel/Coffee.swift:
```diff
@@ -8,19 +8,18 @@
-        case TOP
-        case 칸타타
-        case 조지아
+    public enum CoffeeBean {
```
이런 경우 Coffee.CoffeeBean 타입이 됩니다.
앞에 이름이 네임스페이스 역할을 하기 때문에 Coffee.Bean 형태도 매우 좋습니다.

JUNGYUN-Daegu commented:
> > ## 주요 작업 목록
> > 
> > * [x]   자판기 구조체 설계
> > * [x]   자판기 금액을 원하는 금액만큼 올리는 메소드
> > * [x]   특정 상품 인스턴스를 넘겨서 재고를 추가하는 메소드
> > * [x]   현재 금액으로 구매가능한 음료수 목록을 리턴하는 메소드
> > * [x]   음료수를 구매하는 메소드
> > * [x]   잔액을 확인하는 메소드
> > * [ ]   전체 상품 재고를 (사전으로 표현하는) 종류별로 리턴하는 메소드
> > * [x]   유통기한이 지난 재고만 리턴하는 메소드
> > * [ ]   따뜻한 음료만 리턴하는 메소드
> > * [ ]   시작이후 구매 상품 이력을 배열로 리턴하는 메소드
> > * [ ]   각 메소드 동작을 검증할 수 있는 테스트 함수를 작성
> > * [x]   전체 코드를 확인할 수 있는 통합 테스트 시나리오를 가지고 동작을 확인
> > * [x]   상-하위 클래스 구현 및 이니셜라이저 활용
> > * [x]   클래스에 프로토콜 적용. 동일 프로토콜 채용한 객체간 통신
> > * [ ]   클로저와 핸들러를 이용한 객체간 통신 공부와 해당 메소드 구현
> > * [ ]    XCTest 활용
> > * [x]   Date 자료형 비교
> 
> 완료가 안된 기능은 일부러 빠드린 걸까요?

godrm commented on VendingMachineApp/VendingMachineApp/BeverageModel/CoffeeModel/CoffeeStyleModel/Americano.swift:
```diff
@@ -0,0 +1,21 @@
+    private let isHighCaffeine: Bool
+    
+    init(brandName: String, sizeInMilliliter: Int, itemPrice: Int, itemName: String, manufacturingDate: Date, coffeeBean: CoffeeBean, caffeine: Int) {
+        isHighCaffeine = (caffeine > 150)
```
이런 로직을 초기화할 때 지정하는 것보다는
caffeine 함량을 속성으로 가지고, isHighCaffeine() 같은 메소드를 제공하는 게 좋습니다.

godrm commented on VendingMachineApp/VendingMachineApp/BeverageModel/CoffeeModel/CoffeeStyleModel/Americano.swift:
```diff
@@ -0,0 +1,21 @@
+        super.init(brandName: brandName, sizeInMilliliter: sizeInMilliliter, itemPrice: itemPrice, itemName: itemName, manufacturingDate: manufacturingDate, coffeeBean: coffeeBean)
+    }
+    
+    convenience init(brandName: String, sizeInMilliliter: Int, itemPrice: Int, itemName: String, manufacturingDate: Date, caffeine: Int) {
```
convenience init()에 대해 학습해보세요. 왜 기본 init()과 형태가 동일한 것 같죠?

jung-yun commented:
따뜻한 음료만 리턴하는 메소드 - 질문거리의 인터페이스를 적용하는 방법에 대해 피드백 받으면 구현하려고 남겨두었습니다. 오늘 수업내용에서 어느정도 해결이 되었어요. 
 시작이후 구매 상품 이력을 배열로 리턴하는 메소드 - 구현하려다 막혀서 구현하지않고 넘어갔습니다.

위 사항들과 여타 사항들은 어떻게 구현해야하는지 감을 못잡고 있다가 피드백에서 힌트를 얻고 구현하면 어떨까 생각하게되었습니다.

godrm commented on VendingMachineApp/VendingMachineApp/BeverageModel/CoffeeModel/CoffeeStyleModel/CaffeLatte.swift:
```diff
@@ -0,0 +1,27 @@
+
+class CaffeLatte: Coffee {
+    private let withSteam: Bool
+    private let withSugar: Bool
```
Bool 타입은 매우 정보가 단순합니다. 이런 값보다는 데이터 모델을 할 때는 어떤 데이터를 보관하는지 고민해보세요.

godrm commented on VendingMachineApp/VendingMachineApp/BeverageModel/Drinkable.swift:
```diff
@@ -0,0 +1,12 @@
+
+import Foundation
+
+protocol Drinkable {
+    func isExpired(by date: Date) -> Bool
+}
```
이런 추상화한 프로토콜은 매우 좋습니다. 
다만 어느 타입에서 어떤 용도로 사용하려는 건지 명확하면 더 좋습니다. 
다른 곳에서 채택하는 곳이 전혀 없다면 오히려 없어도 되나 생각해보세요.

godrm commented on VendingMachineApp/VendingMachineApp/BeverageModel/MilkModel/Milk.swift:
```diff
@@ -0,0 +1,25 @@
+import Foundation
+
+class Milk: Beverage {
+    public enum FarmLocation {
+        case domesticFarm
+        case overseasFarm
+    }
+    private let isFrom: FarmLocation
```
이런 시도는 매우 좋지만 말한 것처럼 enum 타입으로 제한을 두는게 오히려 변화를 제약하는 게 아닌가 고민해봐야 합니다.

godrm commented on VendingMachineApp/VendingMachineApp/BeverageModel/SodaModel/Soda.swift:
```diff
@@ -8,19 +8,14 @@
     
-    convenience init(brandName: String, sizeInMilliliter: Int, itemPrice: Int, itemName: String, sodaCategory: SodaCategory) {
-        self.init(brandName: brandName, sizeInMilliliter: sizeInMilliliter, itemPrice: itemPrice, itemName: itemName, manufacturingDate: Date(), sodaCategory: sodaCategory)
+    convenience override init(brandName: String, sizeInMilliliter: Int, itemPrice: Int, itemName: String, manufacturingDate: Date) {
+        self.init(brandName: brandName, sizeInMilliliter: sizeInMilliliter, itemPrice: itemPrice, itemName: itemName, manufacturingDate: manufacturingDate, isZeroCalories: false)
     }
```
속성이 많아지면 생성도 고민을 해야 하는데요, 팩토리 패턴에 대해 학습해보고
음료수 객체들 생성하는 팩토리를 구현해보세요.

godrm commented on VendingMachineApp/VendingMachineApp/BeverageModel/SodaModel/SodaTasteModel/Cola.swift:
```diff
@@ -0,0 +1,26 @@
+    public enum BottleType {
+        case 알루미늄
+        case 유리
+        case 플라스틱
```
코드 케이스에 한글을 쓰면 어떤 장점이 있고, 어떤 단점이 있을까요?

godrm commented on VendingMachineApp/VendingMachineApp/BeverageModel/SodaModel/SodaTasteModel/Fanta.swift:
```diff
@@ -0,0 +1,21 @@
+import Foundation
+
+class Fanta: Soda {
+    private let isFantastic: Bool
```
이 속성이 있는 건 알겠는데, 이게 어떻게 사용되야 할까요. 메소드가 있을까요?

godrm commented on VendingMachineApp/VendingMachineApp/Item.swift:
```diff
@@ -0,0 +1,69 @@
+        self.amount -= amount
+    }
+    
+    public func isSameItem(with item: Item) -> Bool {
+        if self.beverage.isSameBeverage(with: item.beverage) {
+            return true
+        } else {
+            return false
+        }
+    }
+    
+    public func isSoldOut() -> Bool {
+        self.amount == 0 ? true : false
+    }
+    
+    public func isPurchased(by amount: Int = 1) -> Beverage {
```
Beverage와 Item 이라는 타입의 관계를 정하고,
이걸 사용하는 상위 타입에서는 어떤 것을 써야할 지 명확하게 정하는 게 좋습니다.
둘 다 쓰는 것보다는 둘 중에 하나만 사용하는 게 좋으니까요.
매개변수나 리턴 타입에서도 둘 중에 하나만 사용하는 게 좋습니다.

godrm commented on VendingMachineApp/VendingMachineApp/CashInteractor.swift:
```diff
@@ -0,0 +1,53 @@
+
+class CashInteractor: MoneyExchangable {
+    var money: Int
+    private var balance: Int
```
아래보니 MoneyStorage 라는 타입이 있던데 CashInteractor와 역할 분배가 어떻게 되는지 모호합니다.
두 개가 역할과 책임을 나눠갖고 있는 느낌인데 어떤가요?

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -8,18 +8,66 @@
+        machine.showPurchasables()
+        machine.transactionStopButtonPressed()
+        machine.showMoneyStorageDeposit()
+        print("거래 끝")
```
거래가 끝난 것도 machine에서 관리하면 되는거 아닐까요?
machine에서 출력을 할 필요는 없습니다. machine 상태와 결과에 따라서, 출력하는 타입에 전달하는 구조가 좋습니다. 
출력하는 부분은 Machine이랑 분리해보세요.

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -8,18 +8,66 @@
+        let userCalendar = Calendar(identifier: .gregorian)
+        guard let someDateTime = userCalendar.date(from: dateComponents) else { return Date(timeIntervalSinceReferenceDate: -123456789.0)}
+        
+        return someDateTime
```
이 메소드는 ViewController와 전혀 상관이 없는 것 같은데요?

godrm reviewed (changes requested):
다음 피드백을 확인해보세요.

jung-yun committed:
Add DateExtension file

Added DateExtension file creating mother directory for supporting files
Moved Date extension to the new file
Edited useSimpleFormat method more flexible by allowing to receive an argument

jung-yun committed:
Edit isExpired() method

Edited isExpired() method to be more intuitive by delaring constants in it

jung-yun committed:
Extend Equatable protocol in beverage class

And deleted isSameBeverage() method which was predecessor
Amended isSameItem method in Item class
Added another scenario to check the code

jung-yun committed:
Change parameter names at Beverage class's initializer

jung-yun committed:
Add ExpressibleByIntegerLiteral on customized Milli_Liter type

jung-yun committed:
Make farm property in Milk class more flexible

Instead using nested enum, tried to use LocationTrackable protocol as the data of farm
Classes using LocationTrackable are DomesticLocation, OversaesLocaiton classes.
farm property has a flexibility to use both of those classes
Changed relevant initializers
Applied FoodColoringApplicable to the third level classes

jung-yun committed:
Changed Korean variables into English

jung-yun committed:
Implement FoodColoring class

Created FoodColoring class and apply this data type into FoodColoringApplicable protocol
Applied FoodColoringApplicable protocol to Fanta class as well

jung-yun committed:
Delete Drinkable protocol

Deleted drinkable protocol since the only mothod isExpired() can simply be planted in Beverage class

jung-yun committed:
Create HotBeverageAvailable protocol

Applied the protocol to Subclasses of Coffee class
Combined CaffeineSuppliable and HotBeverageAvailable protocol into EnvergyBoostable protocol
Combined CaffeineSuppliable and FoodColoringApplicable into IsHazardous protocol and applied it on Fanta class

jung-yun committed:
Re-construct money handling logic

Combined previous CashInteractor class and MoneyStorage class into MoneyProcessingUnit class
I reckon combining those classes more desiable in terms of abstarction principle of OOP
And this change made code more clean

jung-yun committed:
Apply hashable protocol to Beverage

jung-yun committed:
Improve beverageStorage handling logic

Deleted Item class since Item class and BeverageStorage class had similar functionalities making redundant code on both sides
Made previous Item class into a dictionary
Updated relevant methods

jung-yun committed:
Implement XCTest

Created XCTest file and BeverageUnitTest class to validate Beverage class
Created BeverageFactory to create needed instance of a specific class

jung-yun committed:
Implement unit test for Machine class

jung-yun committed:
Delete dummy function

jung-yun committed:
Implement purchase history showing method

jung-yun commented:
### 부족했던 기능들을 마저 구현했습니다.
- [x] 전체 상품 재고를 (사전으로 표현하는) 종류별로 리턴하는 메소드
- [x] 따뜻한 음료만 리턴하는 메소드
- [x] 시작이후 구매 상품 이력을 배열로 리턴하는 메소드
- [x] 각 메소드 동작을 검증할 수 있는 테스트 함수를 작성
- [x] 클로저와 핸들러를 이용한 객체간 통신 공부와 해당 메소드 구현
- [x]  XCTest 활용

godrm commented on VendingMachineApp/VendingMachineApp/BeverageModel/Beverage.swift:
```diff
@@ -9,56 +9,57 @@ import Foundation
-        } else {
-            return false
-        }
+        let result = self.price <= money ? true : false
+        return result
```
이정도는 그냥 `return self.price <= money` 로 표현해도 됩니다.

godrm commented on VendingMachineApp/VendingMachineApp/BeverageModel/Beverage.swift:
```diff
@@ -9,56 +9,57 @@ import Foundation
     
     public func checkPrice() -> Int {
         return self.price
     }
```
checkPrice는 뭘한다는 것인지 불명확해보입니다. 
개인적으로 check-는 모호한 표현 같아요.

jung-yun committed:
Edit purchase history showing logic and add test case

godrm commented on VendingMachineApp/VendingMachineApp/BeverageModel/Beverage.swift:
```diff
@@ -9,56 +9,57 @@ import Foundation
     }
+    
+    public func isExpired() -> Bool {
+        let oneDayInSeconds: Double = 86400
+        let expiringDateFromPackDate: Double = 60
+        return Date() >= packDate.addingTimeInterval(oneDayInSeconds * expiringDateFromPackDate)
```
여기 쓰는 상수와 메소드를 Date에 확장해서 packDate만 넘기면 처리해주는 게 어떨까요? 
전형적인 상위 모듈 - 하위 모듈 구조라 개선이 가능할 것 같습니다.

godrm commented on VendingMachineApp/VendingMachineApp/BeverageModel/Beverage.swift:
```diff
@@ -9,56 +9,57 @@ import Foundation
 class Beverage {
     private let brand: String
-    private let size: Int
+    private let size: Milli_Liter
```
타입 이름에는 `_` 언더바를 잘 안쓰는 편입니다.

godrm commented on VendingMachineApp/VendingMachineApp/BeverageModel/Beverage.swift:
```diff
@@ -9,56 +9,57 @@ import Foundation
+    private let size: Milli_Liter
     private let price: Int
     private let name: String
     private let packDate: Date
```
속성 이름에 Date 타입 이름을 붙이지 마세요.

godrm commented on VendingMachineApp/VendingMachineApp/BeverageModel/CoffeeModel/CoffeeStyleModel/CaffeLatte.swift:
```diff
@@ -7,21 +7,25 @@
-    private let withSteam: Bool
-    private let withSugar: Bool
+class CaffeLatte: Coffee & EnergyBoostable {
+    var inHeatingCabinet: Bool
```
이건 음료 내부 속성이 아닌게 아닐까요?
예를 들면 HeatingCabinet에 특정 CaffeLatte 인스턴스가 포함되는 방식이 어때요?

godrm commented on VendingMachineApp/VendingMachineApp/BeverageModel/SodaModel/Soda.swift:
```diff
@@ -10,12 +10,12 @@ import Foundation
 class Soda: Beverage {
     private var zeroCalories: Bool
```
zeroCalories 이 속성도 calorie를 용량으로 표현하고 메소드르로 제공하는 게 좋을 것 같습니다.
속성들은 Bool 타입인지 주의할 필요가 있습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Machine.swift:
```diff
@@ -8,61 +8,61 @@
 
 struct Machine {
-    private var moneyStorage = MoneyStorage()
+    private var moneyProccesor = MoneyProcessingUnit()
```
MoneyProcessingUnit 이름이 엄청 거창하네요 ㅎㅎ 
나쁘다는 뜻은 아니구요 재밌습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Machine.swift:
```diff
@@ -8,61 +8,61 @@
+        do {
+            try moneyProccesor.deductMoneyOnTransaction(by: itemPrice)
+        } catch {
+            return
+        }
+        moneyProccesor.increaseHolding(by: itemPrice)
+        do {
+            try beverageStorage.decreaseStock(beverage: beverage)
+        } catch {
+            print("Invalid beverage info")
```
여기서 두 번의 catch가 가능한데요, 하나는 그냥 return 하나는 print만 합니다.
이걸 어떻게 구분할 수 있나요? completionHandler가 있으면 실패할 때도 호출해주는 게 어떨까요?

godrm commented on VendingMachineApp/VendingMachineApp/MoneyProcessingUnit.swift:
```diff
@@ -0,0 +1,53 @@
+    private var moneyOnTransaction: Int
+    
+    init() {
+        self.holding = 0
+        self.moneyOnTransaction = 0
+    }
+    
+    public func increaseHolding(by amount: Int) {
+        holding += amount
+    }
+    
+    public func increaseMoneyOnTransaction(by amount: Int) {
+        moneyOnTransaction += amount
+    }
```
속성이 두 개고 메소드도 2개씩 있는데 어떤 차이가 있는지 모르겠습니다. 
어떻게 구분해서 사용해야 할까요? 하나만 사용할 수는 없나요?

godrm commented on VendingMachineApp/VendingMachineApp/SupportingFiles/GlobalFunctions.swift:
```diff
@@ -0,0 +1,21 @@
+
+import Foundation
+
+func generateDate(year newYear: Int, month newMonth: Int, day newDay: Int) -> Date {
+    var dateComponents = DateComponents()
+    dateComponents.year = newYear
+    dateComponents.month = newMonth
+    dateComponents.day = newDay
+    dateComponents.timeZone = TimeZone(abbreviation: "JST")
+    
+    let userCalendar = Calendar(identifier: .gregorian)
+    guard let someDateTime = userCalendar.date(from: dateComponents) else { return Date(timeIntervalSinceReferenceDate: -123456789.0)}
+    
+    return someDateTime
+}
```
이 글로벌 함수는 뭔가요? 관련있는 타입에 소속을 만들어주세요. 
가급적 글로벌 함수는 만들지 마세요.

godrm commented on VendingMachineApp/VendingMachineApp/SupportingFiles/Milli_Liter.swift:
```diff
@@ -0,0 +1,29 @@
+import Foundation
+
+struct Milli_Liter: ExpressibleByIntegerLiteral {
+    private let value: Int
```
세부 타입을 선언한 것은 매우 좋습니다만 그 역할이 매우 빈약하네요. 
상수값 정도를 포함해도 좋을 것 같습니다.
